src/main.rs

```diff
@@ -99,28 +128,52 @@ fn main() -> Result<()> {
         info!("Battery status: {:?}", &batteries);
 
         let global = system::get_global_battery(&batteries);
+        let battery_state_name = system::battery_state_to_name(global.state);
+
+        if battery_state_name != last_battery_state_name {
+            last_battery_state_name = battery_state_name.clone();
+
+            match battery_state_name.as_str() {
+                "Charging" => run_command(&cfg.battery_state_charging_command),
+                "Discharging" => run_command(&cfg.battery_state_discharging_command),
+                "Not charging" => run_command(&cfg.battery_state_not_charging_command),
+                "Full" => run_command(&cfg.battery_state_full_command),
+                "Unknown" => run_command(&cfg.battery_state_unknown_command),
+                "At threshold" => run_command(&cfg.battery_state_at_threshold_command),
+                _ => {}
+            }
+        }
+
         info!("Global status: {:?}", &global);
-        state_notif.show(
-            format!(
-                "Battery now {}",
-                system::battery_state_to_name(global.state).to_lowercase()
-            ),
-            Urgency::Normal,
-        );
+        if cfg.state_notif_enabled {
+            state_notif.show(
+                format!("Battery now {}", battery_state_name.to_lowercase()),
+                Urgency::Normal,
+                cfg.state_notif_timeout,
+            );
+        }
 
         let level = global.level();
 
         if global.state == system::BatteryState::Charging || level > cfg.low_pct {
             low_notif.close();
         } else if level <= cfg.sleep_pct {
-            low_notif.show("Battery critical".to_string(), Urgency::Critical);
+            low_notif.show(
+                "Battery critical".to_string(),
+                Urgency::Critical,
+                cfg.sleep_pct_notif_timeout,
+            );
             // Just in case we've gone loco, don't do this more than once a minute
             if start > next_sleep_epoch {
                 next_sleep_epoch = start + sleep_backoff;
-                run_sleep_command(&cfg.sleep_command);
+                run_command(&cfg.sleep_command);
             }
         } else if level <= cfg.low_pct {
-            low_notif.show("Battery low".to_string(), Urgency::Critical);
+            low_notif.show(
+                "Battery low".to_string(),
+                Urgency::Critical,
+                cfg.low_pct_notif_timeout,
```

Comment:
If you use an enum, you can map all of this information: the string, the criticality, and the timeout, and just do `low_notif.show()`, and similarly elsewhere. Then we don't have to reimplement for every type of notification.

Reply:
Can you explain a bit more please? Thanks.

Reply:
I'm suggesting using a tagged union and struct to store some of the repeated info. :-)

But I must also confess I don't understand the point of having a different timeout for every single notification type. What's the use case? Urgency::Critical notifications almost always are not dismissed anyway.

Overriding this by the application should generally not be done, the user sets their own preferences in the notification daemon. So unless there is a compelling reason, the timeout stuff should be configured there, not here.

Reply:
> I'm suggesting using a tagged union and struct to store some of the repeated info. :-)

I think I might have done just that already, this is a bit outdated. If not, let me know what improvement can be made. Thanks.

> But I must also confess I don't understand the point of having a different timeout for every single notification type. What's the use case? Urgency::Critical notifications almost always are not dismissed anyway.

I am using [mako](https://github.com/emersion/mako), a very popular server, and it doesn't have per-urgency timeouts.

> Overriding this by the application should generally not be done, the user sets their own preferences in the notification daemon. So unless there is a compelling reason, the timeout stuff should be configured there, not here.

So this is now more of allowing the user to enable/disable individual notifications, while having some extra flexibility while at it. `notify-rust` uses the same integer timeout parameter for making notifications persistent/default timeout/configurable timeout.

Wouldn't applications be disallowed to select this stuff by their choosing if it wasn't meant to be so? Also, it gets a bit tedious at best and very hacky at worst to configure timeout for each notification separately in the notification server.